### PR TITLE
feat(smb2): Add truncate method

### PR DESCRIPTION
**Pull request by GHEMID-Mohamed**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -217,6 +217,22 @@ smb2Client.writeFile('path\\to\\my\\file.txt', 'Hello Node', function(err) {
 });
 ```
 
+> `smb2Client.truncate ( filename, length, [options], callback )`
```
Comment (julien-f) [inline README.md]: What is `options`?

File: README.md [hunk 2]
```diff
@@ -217,6 +217,22 @@ smb2Client.writeFile('path\\to\\my\\file.txt', 'Hello Node', function(err) {
...
+
+-   `filename` String
```
Comment (julien-f) [inline README.md:222]: What about `length`?

File: lib/api/truncate.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    SMB2Request(
+      'set_info',
+      {
+        FileId: file.FileId,
+        FileInfoClass: 'FileEndOfFileInformation',
+        Buffer: fileLength.toBuffer(),
+      },
+      connection,
+      function(err) {
+        if (err) cb && cb(err);
```
Comment (julien-f) [inline lib/api/truncate.js]: Explicitly test `err`: `err != null`

Comment (julien-f) [inline lib/api/truncate.js]: Is the file correctly closed when `set_info` fails?

File: lib/api/truncate.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        else fileSizeSetted();
+      }
+    );
+  }
+
+  function closeFile(fileClosed) {
+    SMB2Request('close', file, connection, function(err) {
+      if (err) cb && cb(err);
```
Comment (julien-f) [inline lib/api/truncate.js]: The operation has already succeed, we should probably not report an error if `close` failed.

File: lib/api/truncate.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      else {
+        file = null;
```
Comment (julien-f) [inline lib/api/truncate.js]: Why?

File: README.md (hunk 1 above)
Comment (GHEMID-Mohamed) [reply README.md]: There's no options, my bad :name_badge:

File: lib/api/truncate.js (hunk 5 above)
Comment (GHEMID-Mohamed) [reply lib/api/truncate.js]: Just to clear the file from memory I guess. I've followed what was written here https://github.com/Node-SMB/marsaud-smb2/blob/master/lib/api/writeFile.js#L62

Comment (julien-f) [reply lib/api/truncate.js]: It should not be necessary, JS garbage collector should free this variable as soon as it's no longer accessible.

File: lib/api/truncate.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      {
+        FileId: file.FileId,
+        FileInfoClass: 'FileEndOfFileInformation',
+        Buffer: fileLength.toBuffer(),
+      },
+      connection,
+      function(err) {
+        if (err != null) {
+          cb && cb(err);
+          closeFile();
```
Comment (julien-f) [inline lib/api/truncate.js]: The workflow is too complicated, always call the passed callback, and only this one.

File: lib/api/truncate.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    SMB2Request(
+      'set_info',
+      {
+        FileId: file.FileId,
+        FileInfoClass: 'FileEndOfFileInformation',
+        Buffer: fileLength.toBuffer(),
+      },
+      connection,
+      function(err) {
+        if (err) cb && cb(err);
```
Comment (julien-f) [inline lib/api/truncate.js]: Explicit tests.

File: lib/api/truncate.js [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        else fileSizeSetted();
+      }
+    );
+  }
+
+  function closeFile(fileClosed) {
+    SMB2Request('close', file, connection, function(err) {
```
Comment (julien-f) [inline lib/api/truncate.js]: Explicit tests.

File: lib/api/truncate.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  SMB2Request(
+    'open',
+    { path: path, desiredAccess: FILE_WRITE_DATA },
+    connection,
+    function(err, file) {
+      if (err != null) {
+        return cb(err);
+      }
+
+      setFileSize(file, new BigInt(8, length), connection, function() {
```
Comment (julien-f) [inline lib/api/truncate.js]: `setFileSize` error should be forwarded.

File: lib/api/truncate.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        SMB2Request('close', file, connection, cb);
```
Comment (julien-f) [inline lib/api/truncate.js]: `close` error should not be forwarded.

File: lib/api/truncate.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    { path: path, desiredAccess: FILE_WRITE_DATA },
+    connection,
+    function(err, file) {
+      if (err != null) {
+        return cb(err);
+      }
+
+      setFileSize(file, new BigInt(8, length), connection, function(err) {
+        if (err != null) {
+          return cb(err);
```
Comment (julien-f) [inline lib/api/truncate.js]: The file MUST ALWAYS be closed.

File: lib/api/truncate.js [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        }
+        SMB2Request('close', file, connection, function() {});
```
Comment (julien-f) [inline lib/api/truncate.js]: `cb` MUST ALWAYS be called!

**Pull request merged**